File: jambeth/jambeth-cache-bytecode/src/main/java/de/osthus/ambeth/bytecode/visitor/FieldAndSetterTemplate.java
package de.osthus.ambeth.bytecode.visitor;

import de.osthus.ambeth.bytecode.ClassGenerator;
import de.osthus.ambeth.bytecode.FieldInstance;
import de.osthus.ambeth.bytecode.MethodInstance;
import de.osthus.ambeth.bytecode.behavior.BytecodeBehaviorState;

public class FieldAndSetterTemplate
{
	private final String fieldName;
	private final java.lang.reflect.Method setterMethod;
	private final int fieldAccess;

	public FieldAndSetterTemplate(int fieldAccess, String fieldName, java.lang.reflect.Method setterMethod)
	{
		this.fieldAccess = fieldAccess;
		this.fieldName = fieldName;
		this.setterMethod = setterMethod;
	}

	public FieldInstance getField(ClassGenerator cg)
	{
		FieldInstance f_beanContext = BytecodeBehaviorState.getState().getAlreadyImplementedField(fieldName);
		if (f_beanContext == null)
		{
			f_beanContext = implementSetter(cg);
		}
		return f_beanContext;
	}

	public MethodInstance getSetter(ClassGenerator cg)
	{
		MethodInstance setter = MethodInstance.findByTemplate(setterMethod, true);
		if (setter == null)
		{
			implementSetter(cg);
		}
		return MethodInstance.findByTemplate(setterMethod, false);
	}

	protected FieldInstance implementSetter(ClassGenerator cg)
	{
		String fieldSignature = FieldInstance.getSignatureFromParameterType(setterMethod, 0);
		FieldInstance f_beanContext = new FieldInstance(fieldAccess, fieldName, fieldSignature, setterMethod.getParameterTypes()[0]);
		f_beanContext = cg.implementField(f_beanContext);
		cg.implementSetter(new MethodInstance(setterMethod), f_beanContext);
		return f_beanContext;
	}
}

File: ambeth/Ambeth.Cache.Bytecode/ambeth/bytecode/visitor/FieldAndSetterTemplate.cs
using De.Osthus.Ambeth.Bytecode.Behavior;
using System;
using System.Reflection;

namespace De.Osthus.Ambeth.Bytecode.Visitor
{
    public class FieldAndSetterTemplate
    {
        private readonly String fieldName;
        private readonly MethodInfo setterMethod;
        private readonly FieldAttributes fieldAccess;

        public FieldAndSetterTemplate(FieldAttributes fieldAccess, String fieldName, MethodInfo setterMethod)
        {
            this.fieldAccess = fieldAccess;
            this.fieldName = fieldName;
            this.setterMethod = setterMethod;
        }

        public FieldInstance GetField(IClassVisitor cg)
        {
            FieldInstance f_beanContext = BytecodeBehaviorState.State.GetAlreadyImplementedField(fieldName);
            if (f_beanContext == null)
            {
                f_beanContext = ImplementSetter(cg);
            }
            return f_beanContext;
        }

        public MethodInstance GetSetter(IClassVisitor cg)
        {
            MethodInstance setter = MethodInstance.FindByTemplate(setterMethod, true);
            if (setter == null)
            {
                ImplementSetter(cg);
            }
            return MethodInstance.FindByTemplate(setterMethod, false);
        }

        protected FieldInstance ImplementSetter(IClassVisitor cg)
        {
            FieldInstance f_beanContext = new FieldInstance(fieldAccess, fieldName, setterMethod.GetParameters()[0].ParameterType);
            f_beanContext = cg.ImplementField(f_beanContext);
            cg.ImplementSetter(new MethodInstance(setterMethod), f_beanContext);

            return f_beanContext;
        }
    }
}